dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java

```diff
@@ -239,46 +243,75 @@ public PerHostClientFactory perHost(String serviceName) {
                             resolvedHosts);
 
                     if (targetUris.isEmpty()) {
-                        return ImmutableList.of();
+                        return ImmutableMap.of();
                     }
 
-                    ImmutableList.Builder<DialogueChannel> list = ImmutableList.builder();
+                    Set<InetAddress> selfAddresses;
+                    try {
+                        selfAddresses = NetworkInterface.networkInterfaces()
+                                .flatMap(NetworkInterface::inetAddresses)
+                                .collect(Collectors.toSet());
+                    } catch (SocketException e) {
+                        log.warn("Failed to obtain local addresses from network interfaces", e);
+                        selfAddresses = Set.of();
+                    }
+
+                    ImmutableMap.Builder<PerHostTarget, Channel> map = ImmutableMap.builder();
                     for (int i = 0; i < targetUris.size(); i++) {
                         TargetUri targetUri = targetUris.get(i);
                         ServiceConfiguration singleUriServiceConf = ServiceConfiguration.builder()
                                 .from(serviceConfiguration)
                                 .uris(ImmutableList.of(targetUri.uri()))
                                 .build();
+                        boolean isSelf = targetUri
+                                .resolvedAddress()
+                                .map(selfAddresses::contains)
+                                .orElse(false);
 
                         // subtle gotcha here is that every single one of these has the same channelName,
                         // which means metrics like the QueuedChannel counter will end up being the sum of all of them.
-                        list.add(cache.getNonReloadingChannel(
-                                params,
-                                singleUriServiceConf,
-                                ImmutableList.of(targetUri),
-                                channelName,
-                                OptionalInt.of(i)));
+                        map.put(
+                                new PerHostTarget(targetUri, isSelf),
+                                cache.getNonReloadingChannel(
+                                        params,
+                                        singleUriServiceConf,
+                                        ImmutableList.of(targetUri),
+                                        channelName,
+                                        OptionalInt.of(i)));
                     }
-                    return list.build();
+                    return map.buildKeepingLast();
```

Comment:
I think we want this to fail loudly/catastrophically if something goes wrong

Reply:
Is this safe? There's no guarantee that the target URIs are unique.